src/View/Form/ContextFactory.php

```diff
@@ -0,0 +1,153 @@
+     * @var array
+     */
+    protected $providers = [];
+
+    /**
+     * Constructor.
+     *
+     * @param array $providers Array of provider callables. Each element should
+     *   be of form `['type' => 'a-string', 'callable' => ..]`
+     */
+    public function __construct(array $providers = [])
+    {
+        foreach ($providers as $provider) {
+            $this->addProvider($provider['type'], $provider['callable']);
+        }
+    }
+
+    /**
+     * Create factory instance with providers "array", "form" and "orm".
+     *
+     * @param array $providers Array of provider callables. Each element should
+     *   be of form `['type' => 'a-string', 'callable' => ..]`
+     * @return \Cake\View\Form\ContextFactory
+     */
+    public static function createWithDefaults(array $providers = [])
+    {
+        $providers = [
+            [
+                'type' => 'orm',
+                'callable' => function ($request, $data) {
+                    if (is_array($data['entity']) || $data['entity'] instanceof Traversable) {
+                        $pass = (new Collection($data['entity']))->first() !== null;
+                        if ($pass) {
+                            return new EntityContext($request, $data);
+                        }
+                    }
+                    if ($data['entity'] instanceof EntityInterface) {
+                        return new EntityContext($request, $data);
+                    }
+                    if (is_array($data['entity']) && empty($data['entity']['schema'])) {
+                        return new EntityContext($request, $data);
+                    }
+                }
+            ],
+            [
+                'type' => 'array',
+                'callable' => function ($request, $data) {
+                    if (is_array($data['entity']) && isset($data['entity']['schema'])) {
+                        return new ArrayContext($request, $data['entity']);
+                    }
+                }
+            ],
+            [
+                'type' => 'form',
+                'callable' => function ($request, $data) {
+                    if ($data['entity'] instanceof Form) {
+                        return new FormContext($request, $data);
+                    }
+                }
+            ],
+        ] + $providers;
+
+        return new static($providers);
+    }
+
+    /**
+     * Add a new context type.
+     *
+     * Form context types allow FormHelper to interact with
+     * data providers that come from outside CakePHP. For example
+     * if you wanted to use an alternative ORM like Doctrine you could
+     * create and connect a new context class to allow FormHelper to
+     * read metadata from doctrine.
+     *
+     * @param string $type The type of context. This key
+     *   can be used to overwrite existing providers.
+     * @param callable $check A callable that returns an object
+     *   when the form context is the correct type.
+     * @return void
```

Comment:
Imho generally `add*()` and `set*()` should unless we use return types instead of exceptions for results of the "adding" or "setting".